ntp-client/src/Network/NTP/Client/Query.hs

```diff
@@ -155,20 +155,15 @@ ntpQuery ioManager tracer ntpSettings = do
             traceWith tracer NtpTraceNoLocalAddr
             ioError $ userError "no local address IPv4 and IPv6"
         l -> return l
-    (v4Replies, v6Replies) <- concurrently (runProtocol IPv4 v4LocalAddr v4Servers)
-                                           (runProtocol IPv6 v6LocalAddr v6Servers)
-
-    case v4Replies ++ v6Replies of
-        [] -> do
-            traceWith tracer NtpTraceIPv4IPv6NoReplies
-            return NtpSyncUnavailable
-        l -> case minimumOfThree l of
-            Nothing -> do
-                traceWith tracer NtpTraceReportPolicyQueryFailed
-                return NtpSyncUnavailable
-            Just offset -> do
-                traceWith tracer $ NtpTraceQueryResult $ getNtpOffset offset
-                return $ NtpDrift offset
+    withAsync (runProtocol IPv4 v4LocalAddr v4Servers) $ \ipv4Async ->
```

Comment:
These are all very non-trivial changes.
`waitCatch` catches any exception.
I don't think that is wanted here.
The original version would propagate the exceptions
and these exception were handled in an other place.
As mentioned before, I think using `concurrently` is fine because the threads that are started
`concurrently` use timeouts and catch IOErrors.

Reply:
> `waitCatch` catches any exception.
> I don't think that is wanted here.

This can be changed to avoid caching async exceptions, only other exception that can happen are `IOException`s. Still we log the result, previously we hand at hock error handing at call sites, now its at the top level. I think that's an improvement.

The debatable difference is handling exceptions when we send request.  Now we will retry for all addresses instead of throwing  exception and not sending any more requests on first error.  This has pros and cons, but since we send just a handful of request (e.g. 3), then there's not much cost in doing so.

> concurrently use timeouts and catch IOErrors.

`concurrently` per se has no timeout, but I think you mean that the `runNtpQueries` have timeout (using async exceptions). This will still work the same way, I haven't changed that. We still do `waitAny` on the timeout thread and the receiving thread.

Reply:
Right `concurrently` starts `runProtocol` and `runProtocol` does timeout and exception handling.